app/jobs/import_from_vacancy_source_job.rb

```diff
@@ -0,0 +1,84 @@
+class ImportFromVacancySourceJob < ApplicationJob
+  queue_as :low
+
+  def perform(source_klass)
+    return if DisableExpensiveJobs.enabled?
+
+    @source_name = source_klass.source_name
+    import_from_source(source_klass)
+  end
+
+  private
+
+  def import_from_source(source_klass)
+    total_count = 0
+    source_external_references = []
+    errors = {}
+
+    source_klass.new.each do |vacancy|
+      total_count += 1
+      source_external_references << vacancy.external_reference
+
+      # Avoids unnecessary and expensive DB updates for vacancies that have not changed since they were imported.
+      next if vacancy.persisted? && !vacancy.changed?
+
+      PaperTrail.request(whodunnit: "Import from external source") do
+        # Only attempt to save if there are no errors coming from the source parsing.
+        # If we save a vacancy that had custom errors, they get removed with the saving validation and we will miss
+        # valuable error debugging information.
+        if vacancy.errors.none? && vacancy.save
+          log("Imported vacancy #{vacancy.id}.")
+        else
+          failure = create_failed_imported_vacancy(vacancy)
+          errors[failure.external_reference] = failure.import_errors
+        end
+      end
+    end
+    report_validation_errors(total_count, errors)
+    mark_removed_vacancies_from_source(source_external_references)
+  end
+
+  def mark_removed_vacancies_from_source(source_external_references)
+    Vacancy.live
+           .where(external_source: @source_name)
+           .where.not(external_reference: source_external_references)
+           .update_all(status: :removed_from_external_system)
+  end
```

Comment:
We mark the records as deleted with a single UPDATE query without validations, callbacks or instantiating the objects in memory. It will simply update the state to `:removed_from_exteranl_system` at DB level.